starlark/int.go

```diff
@@ -204,6 +204,11 @@ func (x Int) CompareSameType(op syntax.Token, v Value, depth int) (bool, error)
 func (i Int) Float() Float {
 	iSmall, iBig := i.get()
 	if iBig != nil {
+		// Avoid bad allocation pattern
+		if iBig.BitLen() <= 64 {
+			return Float(iBig.Sign()) * Float(iBig.Uint64())
```

Comment:
> In this very case, float function will allocate first 8 bytes in the tiny chunk (which is 16 bytes). Immediately after it will allocate other 8 bytes in the same chunk (not always as it depends on how much memory left is there in the chunk, but often enough to make it noticeable). The first allocation could be collectable, but since the tiny allocator collects memory only when all the objects in the chunk become unreachable, the entire chunk gets effectively linked to the lifetime of the starlark.Float (as starlark.Value), in practice doubling its "real" size (and keeping alive the nat part of the big.Int)

I don't think that explanation can be correct, because the tiny allocator avoids pinning objects containing pointers that reference potentially larger amounts of memory; see https://github.com/golang/go/blob/go1.10/src/runtime/malloc.go#L639-L641.

I don't doubt that Float.SetInt and Float.Float64 are expensive though, so avoiding them by your change above makes sense in any case. Thanks for the PR.

(BTW, I'm curious what kind of workload you're using Starlark for that exercises floating point numbers. Care to share?)

Reply:
> The big.Int.Uint64 method is not defined for negative values (its doc comment is wrong), so you'll need a case for each sign:
> 
> ```
> if iBig != nil {
>     // Fast path for hardware int-to-float conversions.
>     if iBig.IsUint64() {
>         return Float(iBig.Uint64())
>     } else if iBig.IsInt64() {
>         return Float(iBig.Int64())
>     }
> 
>     f, _ := new(big.Float).SetInt(iBig).Float64()
>     return Float(f)
> }
> ```
> 
> (FWIW, I think this conversion is subtle enough that it belongs in the math/big package, so I've sent https://go-review.git.corp.google.com/c/go/+/453115 as a first step.)

I think this misses some cases (for example when the absolute value is in range [2^63, 2^64) and the sign is negative.

Reply:
> 

The values that are allocated are not `big.Float` (which, as you were saying, have a slice in it and cannot be allocated by the tiny allocator), but the 1-word buffer pointed by its `mant` attribute which, being an `int`, doesn't contain any pointer.

Reply:
> I think this misses some cases (for example when the absolute value is in range [2^63, 2^64) and the sign is negative.

That's true, your code handles those values too, but relies on a currently undocumented assumption. Despite representing half of all "negated uint64" bit patterns, I expect these values are encountered infrequently in practice.

> The values that are allocated are not big.Float (which, as you were saying, have a slice in it and cannot be allocated by the tiny allocator), but the 1-word buffer pointed by its mant attribute which, being an int, doesn't contain any pointer.

> Also, I think that in this case big.Float doesn't escape the function, so it won't be allocated at all (but I'm speculating now, I have no proof)

Correct, the Float does not escape:

$ go build -gcflags=-S ./starlark/ 2>&1 | less
...
go.starlark.net/starlark.Int.Float STEXT size=112 args=0x8 locals=0x48 funcid=0x0 align=0x0
        0x0000 00000 (starlark/int.go:204)        TEXT    go.starlark.n
et/starlark.Int.Float(
...
        0x0018 00024 (starlark/int.go:204)        FUNCDATA        ZR, gclocals·wgcWObbY2HYnK2SU/U22lA==(SB)
        0x0018 00024 (starlark/int.go:204)        FUNCDATA        $1, gclocals·91Vv7Jr8iuzVj8waVJz8oQ==(SB)
        0x0018 00024 (starlark/int.go:204)        FUNCDATA        $2, go.starlark.net/starlark.Int.Float.stkobj(SB)
        0x0018 00024 (starlark/int.go:204)        FUNCDATA        $5, go.starlark.net/starlark.Int.Float.arginfo1(SB)
        0x0018 00024 (starlark/int.go:204)        FUNCDATA        $6, go.starlark.net/starlark.Int.Float.argliveinfo(SB)
        0x0018 00024 (starlark/int.go:204)        PCDATA  $3, $1
        0x0018 00024 (starlark/int.go:205)        PCDATA  $1, $1
        0x0018 00024 (starlark/int.go:205)        CALL    go.starlark.net/starlark.Int.get(SB)
        0x001c 00028 (starlark/int.go:206)        CBZ     R1, 68
        0x0020 00032 (starlark/int.go:207)        STP     (ZR, ZR), go.starlark.net/starlark..autotmp_10-40(SP)
        0x0024 00036 (starlark/int.go:207)        STP     (ZR, ZR), go.starlark.net/starlark..autotmp_10-24(SP)
        0x0028 00040 (starlark/int.go:207)        MOVD    ZR, go.starlark.net/starlark..autotmp_10-8(SP)
        0x002c 00044 (starlark/int.go:207)        MOVD    $go.starlark.net/starlark..autotmp_10-40(SP), R0
        0x0030 00048 (starlark/int.go:207)        CALL    math/big.(*Float).SetInt(SB)
        0x0034 00052 (starlark/int.go:207)        CALL    math/big.(*Float).Float64(SB)
...```

So I don't see how it can be the cause. My guess is that what you are seeing is the allocation behind z.mant.set in Float.SetInt, which does call runtime.makeslice. In short, the big.Float code is just slow and your optimization avoids it.

Reply:
Sorry, I am slow today. I think I now see what you are trying to tell me: although the Float does not escape, the z.mant.set call allocates a small array that does escape, and this array is tiny-allocated, multiplying its cost from a single word to a whole chunk.

Reply:
That code is indeed "slow" compared to fixed-size manipulation, but the problem lies in `SetInt` (as you were saying) which calls `nat.set` which in turn call `nat.make`. When its argument is 1 (like in this case) it will calls directly `make(nat, 1)`  [here](https://github.com/golang/go/blob/master/src/math/big/nat.go#L67) which will escape. This will cause an allocation of 8 bytes (in the tiny allocator) which is what is causing the problem.

Regarding my implementation, you are right, it is relying on an implementation detail. If I used `big.Int.Bits()` function instead? It's documentation states:

```text
Bits provides raw (unchecked but fast) access to x by returning its absolute value as a little-endian Word slice.
```
So it could be better, what do you think?

Reply:
I'd rather not use Bits() as getting the logic right on a 32-bit machine is more fiddly than I suspect the optimization is worth. Let's stick to the IsInt64+Uint64 variant if you don't mind.

Reply:
I don't mind sticking to your version. Just one last thought: I wouldn't need to cope with little-endian representation as the only case that exhibits this behavior is when the size of the slice is `1`. In all other cases the `nat.make` logic will allocate at least 5 words (e.g. this problem never arises in 32-bit platforms as the only case in which it could happen is not using `big.Int` but `int` directly)

Reply:
Thanks for the discussion so far BTW.

Do you have some guidance on some benchmark/testing I can write for this PR? :-)

Reply:
I changed the code to use the safer alternative, as you proposed.

Reply:
You could add a Starlark benchmark function for this specific operation but the code is so straightforward and obviously more efficient that I don't think it's necessary.